fastscanner/services/indicators/service.py

```diff
@@ -136,18 +138,90 @@ async def handle(self, channel_id: str, data: dict[Any, Any]) -> None:
             if "timestamp" not in data:
                 logger.warning(f"Missing timestamp in message from {channel_id}")
                 return
+
             ts = pd.to_datetime(int(data["timestamp"]), unit="ms", utc=True).tz_convert(
                 LOCAL_TIMEZONE_STR
             )
             new_row = pd.Series(data, name=ts)
 
-            for indicator in self._indicators:
-                new_row = await indicator.extend_realtime(self._symbol, new_row)
-
-            self._handler.handle(self._symbol, new_row)
+            await self._buffer.add(new_row)
 
         except Exception as e:
             logger.error(
                 f"[Handler Error] Failed processing message from {channel_id}: {e}"
             )
+
+
+class CandleBuffer:
+    def __init__(
+        self,
+        symbol: str,
+        freq: str,
+        handler: SubscriptionHandler,
+        indicators: list[Indicator],
+    ):
+        self.symbol = symbol
+        self.freq = freq
+        self.handler = handler
+        self.indicators = indicators
+        self.buffer: dict[pd.Timestamp, pd.Series] = {}
+        self.start_time: pd.Timestamp | None = None
+        self.expected_count = self._get_expected_count()
+        self.timeout_task: asyncio.Task | None = None
```

Comment:
We can implement this logic in the CandleChannelHandler. It seems to me like the candle aggregation it's the handler's responsibility.

Reply:
When you move this to the CandleChannelHandler, please get use to defining most of the variables private (starting with underscore) to maintain the integrity of an object from within the object. Unless there is a strong reason to leave them public.